### ISSUE TITLE
random timeouts

### ISSUE THREAD

**Issue by sishrad** [lusum/ampache-xbmc-plugin#9]
TL;DR This most often occurs at the end of a song but has occurred when launching the plugin. During playback, the playlist proceeds to the next song once the timeout period has elapsed and the error notification is shown.

Environment: RaspPi Zero W running Kodi 17.6 and ampache-xbmc-plugin 0.8.0.2

Everything was running fine (except for the minor scrobble issue I raised last year). However when I recently upgraded Kodi to 18.4 I encountered frequent crashes. So I wiped the memory card and reinstalled 17.6. Then this issue began. I cannot think of how this install differs from the original.

I have run update on the server and restarted it, to no avail (I noticed it updated some Python libraries).

From kodi.log:
16:05:12.178 T:1543500704   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <class 'socket.error'>
                                            Error Contents: [Errno 104] Connection reset by peer
                                            Traceback (most recent call last):
                                              File "/storage/.kodi/addons/plugin.audio.ampache/default.py", line 695, in <module>
                                                play_track(object_id)
                                              File "/storage/.kodi/addons/plugin.audio.ampache/default.py", line 201, in play_track
                                                elem = ampache_http_request("song",filter=id)
                                              File "/storage/.kodi/addons/plugin.audio.ampache/default.py", line 352, in ampache_http_request
                                                response = urllib2.urlopen(req, timeout=400)
                                              File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
                                              File "/usr/lib/python2.7/urllib2.py", line 429, in open
                                              File "/usr/lib/python2.7/urllib2.py", line 447, in _open
                                              File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
                                              File "/usr/lib/python2.7/urllib2.py", line 1228, in http_open
                                              File "/usr/lib/python2.7/urllib2.py", line 1201, in do_open
                                              File "/usr/lib/python2.7/httplib.py", line 1121, in getresponse
                                              File "/usr/lib/python2.7/httplib.py", line 438, in begin
                                              File "/usr/lib/python2.7/httplib.py", line 394, in _read_status
                                              File "/usr/lib/python2.7/socket.py", line 480, in readline
                                            error: [Errno 104] Connection reset by peer
                                            -->End of Python script error report<--
16:05:12.389 T:1630532512 WARNING: CreateLoader - unsupported protocol(plugin) in plugin://plugin.audio.ampache/?mode=9&object_id=72448
16:05:12.390 T:1630532512   ERROR: Init: Error opening file plugin://plugin.audio.ampache/?mode=9&object_id=72448
16:05:12.390 T:1630532512   ERROR: CAudioDecoder: Unable to Init Codec while loading file plugin://plugin.audio.ampache/?mode=9&object_id=72448

**Comment by lusum** [comment]
I have tried the plugin with kodi 18.4 and the current git version, not so different from 0.8.0.2, but i have found no crashes. Can you send me the url you sent to the server? In the log it should be some line before the error. If it doesn't appear, you have to activate notice level in xbmc debug. If the url is correct, it seems an urlib2 error. The function simply send a request to the server for a song. The only problem we could have is incorrect url (  unsupported protocol(plugin) ) . Maybe reinstalling the plugin and check the configuration should help, but without the url i cannot check the corrrectness of the procedure.

**Comment by lusum** [comment]
Ok, i have found that this plugin have some problem with network failures, i will work on it

**Comment by lusum** [comment]
I have committed a better connection error handling, test it if your crashes disappear

**Comment by sishrad** [comment]
Sorry for the delay in response. I've updated to master.zip and am trying to replicate the error. 

Some server requests, if it helps:
10:21:01.966 T:1507382176  NOTICE: http://twomprfm/server/xml.server.php?action=album_songs&auth=925a8bec994a9da459193a46028ca87c&limit=5000&offset=0&filter=4149
10:21:09.230 T:1857033120  NOTICE: http://twomprfm/server/xml.server.php?action=song&auth=925a8bec994a9da459193a46028ca87c&limit=5000&offset=0&filter=49562
10:23:25.870 T:1601880992  NOTICE: http://twomprfm/server/xml.server.php?action=song&auth=925a8bec994a9da459193a46028ca87c&limit=5000&offset=0&filter=49563
10:26:25.761 T:1515770784  NOTICE: http://twomprfm/server/xml.server.php?action=song&auth=925a8bec994a9da459193a46028ca87c&limit=5000&offset=0&filter=49552
10:29:54.135 T:1857033120  NOTICE: http://twomprfm/server/xml.server.php?action=song&auth=925a8bec994a9da459193a46028ca87c&limit=5000&offset=0&filter=49550

I will post any error messages if I encounter any. Thank you.

**Comment by lusum** [comment]
The requests seem regulars

**Comment by lusum** [comment]
I have changed the play_track structure and get rid of the call that cause this problem. The plugin only support 18.4. The crashes are a kodi bug, not an ampache plugin bug, i have found the same bug with other plugin. I explain, if you call another function when a setresolvedurl is called, kodi crashes. So i will wait, as anyone, that kodi team fix this bug.
